### PR TITLE
kubelet: include init containers when determining pod QoS

### LINKED ISSUE #75212: Init containers are not looked at when calculating quality of service class
**What happened**:
A pod that has init container with request for cpu and memory, and containers that make no request for cpu or memory reports a BestEffort QoS class.  This is incorrect.

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/util/qos/qos.go#L35

**What you expected to happen**:
Containers that make requests independent of lifecycle phase should impact quality of service calculation.

Found in 1.12, looks to still be present in 1.14 dev.

/cc  @kubernetes/sig-node-bugs

### PR DESCRIPTION
**Pull request by sjenning**
fixes https://github.com/kubernetes/kubernetes/issues/75212

@derekwaynecarr @dashpole 

```release-note
Init container resource requests now impact pod QoS class
```
/sig node

### CODE REVIEW

**Comment by dashpole** [comment]
This looks correct.
/lgtm

**Comment by dims** [comment]
/assign @liggitt @smarterclayton 

Looks like something we need fixed in 1.14 right? please apply the relevant labels please

File: pkg/apis/core/helper/qos/qos.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ func GetPodQOS(pod *core.Pod) core.PodQOSClass {
 	limits := core.ResourceList{}
 	zeroQuantity := resource.MustParse("0")
 	isGuaranteed := true
-	for _, container := range pod.Spec.Containers {
+	for _, container := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
```
Comment (liggitt) [inline pkg/apis/core/helper/qos/qos.go]: appending to a slice we don't own can end up sharing memory in unpleasant ways. prefer double iterating or appending to a slice we allocated (didn't check how frequently this is called to know which is better)

Comment (sjenning) [reply pkg/apis/core/helper/qos/qos.go]: ok, changed pushed

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/75223#discussion_r264075255 still applies... the changes in this PR are still appending to pod.Spec.Containers

**Comment by sjenning** [comment]
@liggitt can I get `apis/core` approval on this?

**Comment by liggitt** [comment]
This looks reasonable. Do we know what will happen if an existing pod changes QOS class on upgrade as a result of this change? Is this used in the api server or kubelet or both?

**Comment by liggitt** [comment]
/approve

/hold
For question on effect on existing pods on upgrade

**Comment by sjenning** [comment]
The kubelet computes the QoS class whenever it need to know since it already contains the logic.

Some history https://github.com/kubernetes/kubernetes/issues/33255 https://github.com/kubernetes/kubernetes/pull/37968 https://github.com/kubernetes/kubernetes/pull/38989

This change might cause the `QOSClass` field in the Pod `Status` to change.  This does not impact Kubernetes, however, custom clients looking at this field might not expect the change.  `Status` for a resources is understood to be mutable though so... hard to say.  @derekwaynecarr ?

**Comment by liggitt** [comment]
if other places computing this were already doing so considering initContainers, then making this consistent seems reasonable

/hold cancel

**Pull request merged**